### PR TITLE
Authentication Feature

### PR DESCRIPTION
**Pull request by yimingnzhao**
Now requires authentication to be able to post comments, including the following features:
- Comment form and logout button when user is logged in
- Login button when user is logged out
- Backend login checks to prevent html injection of forms
- Changed database structure to include Comment and UserData kinds
- Stores and shows the preferred display name of the current user
- User can change display name, and doing so will change the display name of previously posted comments by user

The current pull request is included with the current deploy, located at:
[https://yimingnzhao-step-2020.appspot.com/](https://yimingnzhao-step-2020.appspot.com/)

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        UserAuth userAuth;
+        UserService userService = UserServiceFactory.getUserService();
+
+        if (userService.isUserLoggedIn()) {
+            String userEmail = userService.getCurrentUser().getEmail();
+            String urlToRedirectToAfterUserLogsOut = "/";
+            String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+            userAuth = new UserAuth(true, "", logoutUrl, userEmail);
```
Comment (jiashenwang) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: Nit: It is a good habit to comment a hard coded value:
userAuth = new UserAuth(/* isLoggedIn= */ true, "", logoutUrl, userEmail);

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        UserAuth userAuth;
+        UserService userService = UserServiceFactory.getUserService();
+
+        if (userService.isUserLoggedIn()) {
+            String userEmail = userService.getCurrentUser().getEmail();
+            String urlToRedirectToAfterUserLogsOut = "/";
```
Comment (jiashenwang) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: Let's make urlToRedirectToAfterUserLogsOut a private static final in LoginServlet

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+            userAuth = new UserAuth(true, "", logoutUrl, userEmail);
+        } else {
+            String urlToRedirectToAfterUserLogsIn = "/";
```
Comment (jiashenwang) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: Let's make the value private static final

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -85,6 +87,13 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
     @Override
     public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        UserService userService = UserServiceFactory.getUserService();
+        if (!userService.isUserLoggedIn()) {
+            response.sendRedirect("/#comments");
+            return;
+        }
+
+        String userEmail = userService.getCurrentUser().getEmail();
         String username = request.getParameter("name");
```
Comment (jiashenwang) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Let's make all the strings here private static final. For example:
private final static PARAM_NAME = "name";

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -93,11 +102,12 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
         Entity commentEntity = new Entity("Comment");
         commentEntity.setProperty("username", username);
         commentEntity.setProperty("message", message);
+        commentEntity.setProperty("email", userEmail);
         commentEntity.setProperty("timestamp", timestamp);
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
         datastore.put(commentEntity);
 
-        response.sendRedirect("/#contact");
+        response.sendRedirect("/#comments");
```
Comment (jiashenwang) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Ditto, make it static final

File: portfolio/src/main/java/com/google/sps/servlets/DeleteCommentsServlet.java [hunk 6]
```diff
@@ -93,7 +94,7 @@ public Long tryParseLong(String str) {
         try {
             return Long.parseLong(str);
         } catch (NumberFormatException e) {
-            return null;
+            return -1;
```
Comment (jiashenwang) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteCommentsServlet.java]: It is better to throw an exception and catch it from the caller

File: portfolio/src/main/java/com/google/sps/servlets/DeleteCommentsServlet.java [hunk 7]
```diff
@@ -90,11 +90,11 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
      * @param str The string to try to parse to a long
      * @return The long value of the string, or -1 if an exception is thrown
      */
-    public Long tryParseLong(String str) {
+    public long tryParseLong(String str) {
         try {
             return Long.parseLong(str);
         } catch (NumberFormatException e) {
-            return -1;
+            return PARSE_LONG_EXCEPTION;
```
Comment (jiashenwang) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteCommentsServlet.java:103]: Great

File: portfolio/src/main/java/com/google/sps/servlets/UserDataServlet.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            response.sendRedirect("/#comments");
+            return;
+        }
+
+        String id = userService.getCurrentUser().getUserId();
+        String email = userService.getCurrentUser().getEmail();
+        String displayName = request.getParameter("name");
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Entity entity = new Entity("UserData", id);
```
Comment (jiashenwang) [inline portfolio/src/main/java/com/google/sps/servlets/UserDataServlet.java]: Ditto, make it a final const

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -170,35 +170,45 @@ function loadComments(query) {
...
+
 /**
  * Edits the comment section based on current login status
  */
 function manageLogin() {
     // Fetches the current login status and changes html based on the status
     fetch('/login').then((response) => response.json()).then((userAuth) => {
         if (userAuth.isLoggedIn) {
-            $('#comments-input').find('form').find('#comment-form-name-input').val(userAuth.userEmail);
+            //$('#comments-input').find('form').find('#comment-form-name-input').val(userAuth.userEmail);
```
Comment (jiashenwang) [inline portfolio/src/main/webapp/script.js]: Remove it if not using

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -223,6 +234,10 @@ function hasOnlyDigits(value) {
     return /^\d+$/.test(value);
 }
 
+function isAlphanumeric(value) {
+    return /^[0-9a-z]+$/.test(value);
```
Comment (jiashenwang) [inline portfolio/src/main/webapp/script.js]: Do we need to consider about capital?

File: portfolio/src/main/java/com/google/sps/servlets/UserDataServlet.java [hunk 11]
```diff
@@ -24,12 +24,45 @@
...
+        }
+
+        String userId = userService.getCurrentUser().getUserId();
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Filter propertyFilter = new FilterPredicate("id", FilterOperator.EQUAL, userId);
+        Query query = new Query("UserData").setFilter(propertyFilter);
+        PreparedQuery results = datastore.prepare(query);
+        List<Entity> listResults = results.asList(FetchOptions.Builder.withDefaults());
+
+        if (listResults.size() != 1) {
```
Comment (jiashenwang) [inline portfolio/src/main/java/com/google/sps/servlets/UserDataServlet.java:75]: Is it possible user id is not unique and we get 2 results?

**Review by jiashenwang — CHANGES_REQUESTED** [review]
Great job. Looks good. I minor issue is that we are using string constants to get values from entity. It will be better if we could centralized all const which could avoid potential bugs.

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 1 above)
Comment (yimingnzhao) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: I made the inputs to isLoggedIn parameter static final booleans

File: portfolio/src/main/webapp/script.js (hunk 10 above)
Comment (yimingnzhao) [reply portfolio/src/main/webapp/script.js]: An updated commit changed the regex to accept capital letters

File: portfolio/src/main/java/com/google/sps/servlets/UserDataServlet.java (hunk 11 above)
Comment (yimingnzhao) [reply portfolio/src/main/java/com/google/sps/servlets/UserDataServlet.java:75]: I'm pretty sure user IDs are unique unless something is manually changed on DataStore. I chose to use `listResults.size() != 1`  as the conditional mainly to ensure that the query is not empty, as well as to catch the unexpected case that the user ID is not unique.

**Comment by yimingnzhao** [comment]
The newest commit centralizes all constant Strings as static finals in all java classes.

**Review by jiashenwang — APPROVED** [review]
LGTM. Great job!

**Pull request merged**